src/elements/element.line.js

```diff
@@ -219,6 +221,195 @@ function updateBezierControlPoints(points, options, area, loop) {
+ * @param {boolean} segment.loop - indicates that the segment is a loop
+ * @param {Point[]} points - the points that this segment refers to
+ * @param {object} bounds
+ * @param {string} bounds.property - the property of a `Point` we are bounding. `x`, `y` or `angle`.
+ * @param {number} bounds.start - start value of the property
+ * @param {number} bounds.end - end value of the property
+ **/
+export function _boundSegment(segment, points, bounds) {
+	if (!bounds) {
+		return [segment];
+	}
+
+	const {property, start: startBound, end: endBound} = bounds;
+	const count = points.length;
+	const {compare, between, normalize} = propertyFn(property);
+	const {start, end, loop} = getSegment(segment, points, bounds);
+	const result = [];
+	let inside = false;
+	let subStart = null;
+	let i, value, point, prev;
+
+	for (i = start; i <= end; ++i) {
+		point = points[i % count];
+
+		if (point.skip) {
+			continue;
+		}
+
+		value = normalize(point[property]);
+		inside = between(value, startBound, endBound);
+
+		if (subStart === null && inside) {
+			subStart = i > start && compare(value, startBound) > 0 ? prev : i;
+		}
+
+		if (subStart !== null && (!inside || compare(value, endBound) === 0)) {
+			result.push(makeSubSegment(subStart, i, loop, count));
+			subStart = null;
+		}
+		prev = i;
+	}
+
+	if (subStart !== null) {
+		result.push(makeSubSegment(subStart, end, loop, count));
+	}
+
+	return result;
+}
+
+/**
+ * Returns the segments of then line that are inside given bounds
+ * @param {Line} line
+ * @param {object} bounds
+ * @param {string} bounds.property - the property we are bounding with. `x`, `y` or `angle`.
+ * @param {number} bounds.start - start value of the `property`
+ * @param {number} bounds.end - end value of the `property`
+ */
+export function _boundSegments(line, bounds) {
+	const result = [];
+
+	for (let segment of line.segments) {
+		let sub = _boundSegment(segment, line.points, bounds);
+		if (sub.length) {
+			result.push(...sub);
+		}
+	}
+	return result;
+}
+
+function findStart(points, count, loop) {
+	let start = 0;
+
+	while (loop && start < count && !points[start].skip) {
+		start++;
+	}
+
+	while (start < count && points[start].skip) {
+		start++;
+	}
+
+	return start % count;
+}
+
+/**
+ * Compute the continuous segments that define the whole line
+ * There can be skipped points within a segment, if spanGaps is true.
+ * @param {Line} line
+ */
+function computeSegments(line) {
+	const points = line.points;
+	const spanGaps = line.options.spanGaps;
+	const count = points.length;
+	const result = [];
+	let loop = spanGaps ? !!line._loop : !!line._fullLoop;
+
+	if (!count) {
+		return result;
+	}
+
+	let last = null;
+	let start = findStart(points, count, loop);
+	let max = loop ? count + start : count;
+	let cur, prev, end;
+
+	prev = points[start];
+	if (!prev.skip) {
+		last = start;
+	}
+	for (end = start + 1; end < max; ++end) {
```

Comment:
How does this impact the uPlot benchmark? It looks like we now have to loop over all data points an additional time and I'm worried that might be expensive

Reply:
Inconclusive. uPlot bench uses `fastPath` that skips this segment stuff.

Heres a plunker with some preconfigured uPlot based benchmarks to check it out:
https://plnkr.co/edit/QJ9IMuiybyXWHtgdTfEK?p=preview